Enable support for TI Arm Clang Compiler (tiarmclang)

evodius96 opened the pull request:
- Add new cmsis_tiarmclang.h and modified existing Core header files
- Add linker command files (linker scripts) to supported CoreValidation targets
- Adjust CoreValidation to avoid compiler removal of empty "delay loops"

evodius96 committed:
Enable support for TI Arm Clang Compiler (tiarmclang)
- Add new cmsis_tiarmclang.h and modified existing Core header files
- Add linker command files (linker scripts) to supported CoreValidation targets
- Adjust CoreValidation avoid compiler removal of empty "delay loops"

grasci-arm commented:
Can one of the admins verify this patch?

JonatanAntoni commented on CMSIS/Core/Include/cmsis_tiarmclang.h:
```diff
@@ -0,0 +1,1510 @@
+/**************************************************************************//**
+ * @file     cmsis_tiarmclang.h
+ * @brief    CMSIS compiler tiarmclang header file
+ * @version  V5.6.0
```
This is a new file and version should start at 1.0.0

evodius96 replied:
No problem.   I'm assuming, then, that the file version and date reflect the last file modification?  In what way does the version increment when the file is changed?

JonatanAntoni replied:
Yes, reason behind having individual file versions and dates is that still people are used to ship copies of cmsis files (instead of relying on official releases of the pack). In order to sort out user feedback we need to specifically know which version if a files a user received. The version shall follow [SemVer](https://semver.org/lang/de/) rules, i.e., increment major part on non-backward compatible changes (should be avoided), minor part for new features, and patch part for bug-fixes only.

JonatanAntoni replied:
Do you plan to set this to 1.0.0 for the initially contributed version?

evodius96 replied:
My second commit makes this change to V1.0.0 -- is that not visible?  If necessary, I can squash the commits.  We use a different review infrastructure, so I apologize if I am not using pull requests optimally.

JonatanAntoni commented on CMSIS/Core/Include/core_armv81mml.h:
```diff
@@ -97,6 +97,29 @@
     #define __DSP_USED         0U
   #endif
 
+#elif defined (__ti__)
```
Adding this for every compiler seems not to be maintainable on the long run when adding more compilers. We need to look for another solution. Can we move this somehow into the compiler header? Or reduce to be compiler independent by adding a define-mapping to compiler headers?

evodius96 replied:
In general, I agree -- there is a lot of duplication here, and there are also some target specific dependencies in a handful of the files (depending on whether the target cares about the compiler enabling FPU instructions, DSP instructions, or not).  Also, tiarmclang also defines __GNUC__, which ties it together.  I'm willing to do what I can for tiarmclang, but I think someone needs to look at the design holistically and then refactor in a way that makes sense.  I think that's beyond the scope of this pull request, although I absolutely understand why you'd call it out.

JonatanAntoni replied:
Yes, agree. We should not do to much at once. I think we eventually can move the defines into the compiler headers. Let's leave it as is for now. I'll put me a note to address this change in CMSIS 6.

JonatanAntoni commented on CMSIS/CoreValidation/Source/CV_CoreFunc.c:
```diff
@@ -32,6 +32,12 @@ static void TC_CoreFunc_IPSR_IRQHandler(void) {
   irqXPSR = __get_xPSR();
 }
 
+#if defined(__ti__)
```
This is only test code, so performance is not big issue.
Hence, I'd add `__NOP` in any case and remove compiler dependency here.

evodius96 replied:
OK will do

JonatanAntoni commented on CMSIS/CoreValidation/Source/CV_CoreFunc.c:
```diff
@@ -307,6 +313,9 @@ void TC_CoreFunc_IPSR (void) {
 #if defined(__CC_ARM)
 #define SUBS(Rd, Rm, Rn) __ASM volatile("SUBS " # Rd ", " # Rm ", " # Rn)
 #define ADDS(Rd, Rm, Rn) __ASM volatile("ADDS " # Rd ", " # Rm ", " # Rn)
+#elif defined(__ti__)
```
Why is this duplication of the else-tree required?
Is it because `__GNUC__` is defined?

JonatanAntoni replied:
Perhaps we need to move these defines into the compiler header?

evodius96 replied:
Yes, `__GNUC__` is also defined.  However, I can just add a` !defined(__ti__)` to the `GNUC `clause and get the same behavior without duplicating the macros.  it's worth exploring moving these defines into the compiler headers.

JonatanAntoni reviewed (changes requested):
Please consider updating all touched files versions and dates

evodius96 committed:
Review updates
- Adjust versions and dates according to semantic versioning
- Use __NOP(); in delay loops rather than compiler-dependent macro
- Remove duplicated ADDS/SUBS macro definitions

JonatanAntoni committed:
Merge branch 'develop' into enable-tiarmclang
